File: user-service/src/utils/logger.js
const amqp = require('amqplib');
const { v4: uuidv4 } = require('uuid');

class Logger {
    constructor(serviceName) {
        this.serviceName = serviceName;
        this.connection = null;
        this.channel = null;
        this.exchange = 'logs_exchange';
        this.queue = 'logging_queue';
        this.initPromise = this.initialize();
    }

    async initialize() {
        try {
            const rabbitmqUrl = process.env.RABBITMQ_URL || 'amqp://admin:admin123@rabbitmq:5672';
            this.connection = await amqp.connect(rabbitmqUrl);
            this.channel = await this.connection.createChannel();

            // Declare exchange
            await this.channel.assertExchange(this.exchange, 'fanout', { durable: true });

            // Declare queue
            await this.channel.assertQueue(this.queue, { durable: true });

            // Bind queue to exchange
            await this.channel.bindQueue(this.queue, this.exchange, '');

            console.log(`Logger initialized for ${this.serviceName}`);

            // Handle connection close
            this.connection.on('close', () => {
                console.error('RabbitMQ connection closed. Reconnecting...');
                setTimeout(() => this.initialize(), 5000);
            });

            this.connection.on('error', (err) => {
                console.error('RabbitMQ connection error:', err.message);
            });

        } catch (error) {
            console.error('Failed to initialize logger:', error.message);
            // Retry connection after 5 seconds
            setTimeout(() => this.initialize(), 5000);
        }
    }

    async log(level, url, correlationId, message, additionalData = {}) {
        try {
            await this.initPromise;

            if (!this.channel) {
                console.error('Logger channel not initialized');
                return;
            }

            const timestamp = new Date().toISOString().replace('T', ' ').substring(0, 23);
            const logMessage = `${timestamp} ${level.toUpperCase()} ${url} Correlation: ${correlationId} [${this.serviceName}] - ${message}`;

            const logData = {
                timestamp,
                level: level.toUpperCase(),
                url,
                correlationId,
                serviceName: this.serviceName,
                message,
                ...additionalData
            };

            this.channel.publish(
                this.exchange,
                '',
                Buffer.from(JSON.stringify(logData)),
                { persistent: true }
            );

            // Also log to console
            console.log(logMessage);
        } catch (error) {
            console.error('Failed to send log to RabbitMQ:', error.message);
            // Fallback to console logging
            console.log(`${level.toUpperCase()} ${url} Correlation: ${correlationId} [${this.serviceName}] - ${message}`);
        }
    }

    info(url, correlationId, message, additionalData) {
        return this.log('INFO', url, correlationId, message, additionalData);
    }

    error(url, correlationId, message, additionalData) {
        return this.log('ERROR', url, correlationId, message, additionalData);
    }

    warn(url, correlationId, message, additionalData) {
        return this.log('WARN', url, correlationId, message, additionalData);
    }

    async close() {
        try {
            if (this.channel) await this.channel.close();
            if (this.connection) await this.connection.close();
        } catch (error) {
            console.error('Error closing logger:', error.message);
        }
    }
}

// Singleton instance
let loggerInstance = null;

function getLogger(serviceName) {
    if (!loggerInstance) {
        loggerInstance = new Logger(serviceName);
    }
    return loggerInstance;
}

// Middleware to add correlation ID
function correlationMiddleware(serviceName) {
    return (req, res, next) => {
        // Get correlation ID from header or generate new one
        req.correlationId = req.headers['x-correlation-id'] || uuidv4();

        // Add correlation ID to response headers
        res.setHeader('X-Correlation-Id', req.correlationId);

        next();
    };
}

// Middleware to log requests
function loggingMiddleware(logger) {
    return (req, res, next) => {
        const startTime = Date.now();
        const url = `${req.protocol}://${req.get('host')}${req.originalUrl}`;
        const correlationId = req.correlationId || 'unknown';

        // Log request
        logger.info(url, correlationId, `${req.method} request received`, {
            method: req.method,
            path: req.path,
            query: req.query,
            ip: req.ip
        });

        // Capture response
        const originalSend = res.send;
        res.send = function (data) {
            res.send = originalSend;

            const duration = Date.now() - startTime;
            const level = res.statusCode >= 400 ? 'error' : 'info';

            logger.log(
                level,
                url,
                correlationId,
                `${req.method} request completed - Status: ${res.statusCode} - Duration: ${duration}ms`,
                {
                    method: req.method,
                    path: req.path,
                    statusCode: res.statusCode,
                    duration
                }
            );

            return res.send(data);
        };

        next();
    };
}

module.exports = {
    Logger,
    getLogger,
    correlationMiddleware,
    loggingMiddleware
};

File: analytics_server/logger.py
import pika
import json
import os
from datetime import datetime
from uuid import uuid4
from flask import request
import threading
import time

class Logger:
    def __init__(self, service_name):
        self.service_name = service_name
        self.connection = None
        self.channel = None
        self.exchange = 'logs_exchange'
        self.queue = 'logging_queue'
        self.rabbitmq_url = os.getenv('RABBITMQ_URL', 'amqp://admin:admin123@rabbitmq:5672')
        self.initialize()

    def initialize(self):
        """Inicializiraj povezavo s RabbitMQ"""
        try:
            # Parse connection URL
            params = pika.URLParameters(self.rabbitmq_url)
            params.socket_timeout = 5
            
            self.connection = pika.BlockingConnection(params)
            self.channel = self.connection.channel()
            
            # Declare exchange
            self.channel.exchange_declare(
                exchange=self.exchange,
                exchange_type='fanout',
                durable=True
            )
            
            # Declare queue
            self.channel.queue_declare(queue=self.queue, durable=True)
            
            # Bind queue to exchange
            self.channel.queue_bind(
                exchange=self.exchange,
                queue=self.queue
            )
            
            print(f"Logger initialized for {self.service_name}")
        except Exception as e:
            print(f"Failed to initialize logger: {str(e)}")
            # Retry after 5 seconds
            threading.Timer(5.0, self.initialize).start()

    def log(self, level, url, correlation_id, message, additional_data=None):
        """Pošlji log v RabbitMQ"""
        try:
            if not self.channel or self.connection.is_closed:
                self.initialize()
            
            timestamp = datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S.%f')[:-3]
            log_message = f"{timestamp} {level.upper()} {url} Correlation: {correlation_id} [{self.service_name}] - {message}"
            
            log_data = {
                'timestamp': timestamp,
                'level': level.upper(),
                'url': url,
                'correlationId': correlation_id,
                'serviceName': self.service_name,
                'message': message
            }
            
            if additional_data:
                log_data.update(additional_data)
            
            self.channel.basic_publish(
                exchange=self.exchange,
                routing_key='',
                body=json.dumps(log_data),
                properties=pika.BasicProperties(
                    delivery_mode=2  # Naredi sporočilo trajno
                )
            )
            
            # Še log na konzolo
            print(log_message)
        except Exception as e:
            print(f"Failed to send log to RabbitMQ: {str(e)}")
            # Povratni log na konzolo
            print(f"{level.upper()} {url} Correlation: {correlation_id} [{self.service_name}] - {message}")

    def info(self, url, correlation_id, message, additional_data=None):
        self.log('INFO', url, correlation_id, message, additional_data)

    def error(self, url, correlation_id, message, additional_data=None):
        self.log('ERROR', url, correlation_id, message, additional_data)

    def warn(self, url, correlation_id, message, additional_data=None):
        self.log('WARN', url, correlation_id, message, additional_data)

    def close(self):
        """Zapri povezavo s RabbitMQ"""
        try:
            if self.channel:
                self.channel.close()
            if self.connection:
                self.connection.close()
        except Exception as e:
            print(f"Error closing logger: {str(e)}")


# Singleton instanca
_logger_instance = None

def get_logger(service_name):
    global _logger_instance
    if _logger_instance is None:
        _logger_instance = Logger(service_name)
    return _logger_instance


def correlation_middleware():
    """Sredstvo za dodajanje correlation ID v zahtevo"""
    correlation_id = request.headers.get('X-Correlation-Id', str(uuid4()))
    request.correlation_id = correlation_id
    return correlation_id


def create_logging_middleware(logger):
    """Ustvari sredstvo za beleženje v Flasku"""
    def logging_middleware():
        from flask import g
        import time
        
        # Get or generate correlation ID
        correlation_id = correlation_middleware()
        g.correlation_id = correlation_id
        
        # Log request
        url = request.url
        start_time = time.time()
        
        logger.info(
            url,
            correlation_id,
            f"{request.method} request received",
            {
                'method': request.method,
                'path': request.path,
                'query': dict(request.args),
                'ip': request.remote_addr
            }
        )
        
        return correlation_id
    
    return logging_middleware


def log_response(logger, correlation_id, url, method, status_code, start_time):
    """Log response"""
    duration = int((time.time() - start_time) * 1000)
    level = 'error' if status_code >= 400 else 'info'
    
    logger.log(
        level,
        url,
        correlation_id,
        f"{method} request completed - Status: {status_code} - Duration: {duration}ms",
        {
            'method': method,
            'statusCode': status_code,
            'duration': duration
        }
    )

